core/src/main/scala/kafka/server/DelayedOperation.scala

```diff
@@ -292,34 +309,46 @@ class DelayedOperationPurgatory[T <: DelayedOperation](purgatoryName: String,
    */
   private class Watchers(val key: Any) {
 
-    private[this] val operations = new LinkedList[T]()
+    private[this] val operations = new ConcurrentLinkedQueue[T]()
+    private[this] val approximateWatched = new AtomicInteger(0)
```

Comment:
approximateWatched can be a bit mis-leading. It's eventually accurate. So, probably just using watched is fine.